bundles/automation/org.eclipse.smarthome.automation.core/src/main/java/org/eclipse/smarthome/automation/core/util/RuleBuilder.java

```diff
@@ -39,14 +39,34 @@
 @NonNullByDefault
 public class RuleBuilder {
 
-    private final RuleImpl rule;
-
-    protected RuleBuilder(RuleImpl rule) {
-        this.rule = rule;
+    private @NonNullByDefault({}) List<Trigger> triggers;
+    private @NonNullByDefault({}) List<Condition> conditions;
+    private @NonNullByDefault({}) List<Action> actions;
+    private @NonNullByDefault({}) Configuration configuration;
+    private @NonNullByDefault({}) List<ConfigDescriptionParameter> configDescriptions;
+    private @Nullable String templateUID;
+    private @NonNullByDefault({}) final String uid;
+    private @Nullable String name;
+    private @NonNullByDefault({}) Set<String> tags;
+    private @NonNullByDefault({}) Visibility visibility;
+    private @Nullable String description;
+
+    protected RuleBuilder(Rule rule) {
+        this.triggers = rule.getTriggers();
```

Comment:
this construct also seems a little bit strange:
I would expect a `Rule` to expose its triggers, conditions, actions, etc only as immutable objects. Here the builders lists are set to the internal lists of a RuleImpl.
Why is a RuleBuilder initialized using a RuleImpl at all? I would expect to builder to holt all the building properties internally and on Rule creation set all the properties accordingly.

Reply:
There is a builder that can be used to "copy" and so modify an existing `Rule` (not only `RuleImpl`).
This has been done in other builder too AFAIK.

> I would expect to builder to holt all the building properties internally and on Rule creation set all the properties accordingly.

Isn't this exactly what I fixed with my PR?

Reply:
agreed on the `copy`mode and constructor. But I would again prefer to hold all the information to build a new Rule inside the builder. Here at least the lists from the RuleImpl are used. I expected to see something like 

```
this.triggers = new ArrayList(rule.getTriggers());
```

Reply:
@htreu Isn't it responsible to the rule implementation to return a copy of its internal data structure?
AFAIK it is a common quality violation if objects return its internal data.

We already create a copy if `getTriggers` of a rule implementation is called, so why copy that copy again?

Reply:
Ah, I see, the `RuleImpl` does not return a copy... Do you agree that we should change the implementation of `RuleImpl` to not expose internal data structures?

Reply:
I should check the code more detailed before writing. `RuleImpl` holds unmodifiable collections and so `RuleImpl` and the builder should be fine.

I still think implementations are responsible to return copies or unmodifiable collections. If you think an additional copy would make sense, okay for me. But I don't think we are more safe and just consume some resources on runtime. :wink:

Reply:
> We already create a copy if getTriggers of a rule implementation is called

From the builder perspective I can not be sure that its always be done this way. The more defensive approach would be to hold internal lists. And for the RuleImpl I would expect the `getTriggers` to expose an immutable list.

Reply:
sorry, I was writing this not seeing your latest comments.

Reply:
Now I get our misunderstanding I guess: I was under the impression that the lists on the builder are used to be further modified. This is actually not (yet) the case.
Nevertheless I would prefer to hold independent list instances on the builder in case we decide to have something like `withTrigger(...)`  methods to add to the list. wdyt?

Reply:
@htreu Sure, we could create our own list to be on the safe side if a rule implementation is used that returns it modifiable internal representation.

Hm, need to check the builder pattern: Do you think `withTriggers` should replace the collection but `withTrigger` should add something to the collection? So, this would mean that I fill a collection calling `withTriggers` containing trigger1 and trigger2 first and then call `withTrigger(trigger3).withTrigger(trigger4)` trigger3 and trigger4 are added. If I call (after the initial fill) `withTriggers(Arrays.asList(trigger3, trigger4))` then trigger1 and trigger2 are not in the trigger list anymore? Is this what you like?
Shouldn't we discuss if there should be an additional `addTrigger` and keep the `withTriggers` as it is?

Reply:
I'm unsure myself about the behaviour of `withTriggers`. But for example in `ThingTypeBuilder#withChannelDefinitions` we are actually overwriting the list (and unfortunately do not make a defensive copy of the elements). And there are other examples like this: `ThingBuilder#withChannels` will overwrite but `ThingBuilder#withChannel` adds to the list.

Reply:
> `ThingBuilder#withChannels` will overwrite but `ThingBuilder#withChannel` adds to the list.

Hm, very confusing situation.
Wouldn't it be more intuitive to use a `with` to set something and an `add` to add something?

Reply:
Yes, agreed.
For this PR: do you consider using internal lists? Since we do not have an `addTrigger` method all should work for the moment.

Reply:
I will use an internal list and change the implementation.

About the `with...s` implementation. I don't think we should ignore `null`. The caller should be able to remove all modules (e.g. trigger, actions, ...). Currently `null` is an allowed argument (annotations) but ignored completely. In our ESH code base we handle null and empty collections most of the time the same way (IIRC) here it is handled differently.

I will try to create an issue (and later a wiki page) about the builder pattern we use in ESH. It should be consistent over all implementations.